emily/handler/src/database/entries/deposit.rs

```diff
@@ -155,6 +156,46 @@ impl DepositEntry {
             self.key(),
         )))
     }
+
+    /// Reorgs around a given chainstate.
+    /// TODO(TBD): Remove duplicate code around deposits and withdrawals if possible.
+    pub fn reorganize_around(&mut self, chainstate: &Chainstate) -> Result<(), Error> {
+        // Update the history to have the histories wiped after the reorg.
+        self.history.retain(|event| {
+            // The event is younger than the reorg...
+            (chainstate.stacks_block_height > event.stacks_block_height)
+                // Or the event is as old as the reorg but has a different block hash...
+                || ((chainstate.stacks_block_height == event.stacks_block_height)
+                    && (chainstate.stacks_block_hash == event.stacks_block_hash))
+        });
+        // If the history is empty add a reassessing event.
+        if self.history.is_empty() {
+            self.history = vec![DepositEvent {
+                status: StatusEntry::Reprocessing,
+                message: "Reprocessing deposit status after reorg.".to_string(),
+                stacks_block_height: chainstate.stacks_block_height,
+                stacks_block_hash: chainstate.stacks_block_hash.clone(),
+            }]
+        }
+        // Synchronize self with the new history.
+        self.synchronize_with_history()?;
+        // Return.
+        Ok(())
+    }
+
+    /// Synchronize the entry with its history.
+    pub fn synchronize_with_history(&mut self) -> Result<(), Error> {
```

Comment:
Can you add a description of why this is necessary? I don't really know what synchronize means here.

Reply:
I'll write it here and also add a comment in the code.

The Deposit and Withdrawal entries both contain an internal vector of history entries in chronological order. The last entry in the history vector is the latest entry, meaning the most up-to-date data. Within this history is the most recent `last_update_height` and `status`.

DynamoDB can only be sorted and indexed by top level fields, so in order to allow the table to be searchable by `status` or ordered by `last_update_height` there needs to be a top level field for it.

This function takes the entry and then synchronizes the top level fields `status` and `last_update_height` with the history vector by taking the latest history event and updating those entry fields to match it. 

I was originally doing this directly a number of places in the code and then I realized it was better to just update the history and have a separate function that synchronizes the top level entry fields with its history.